### PR TITLE
fix: failure in state transition

### LINKED ISSUE #408: failure in state transition
error in bddtest did exchange basic flow

 [aries-framework/transport] 2019/10/03 14:51:16 UTC - http.processPOSTRequest -> ERROR incoming msg processing failed: invalid state transition: requested -> completed

```[aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO send post event for state responded
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO send pre event for state completed
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).Handle -> INFO entered into Handle exchange message : {"@type":"did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/ack","@id":"e189cf47-64f7-45d5-90b1-948ea4c2ba98","status":"ok","~thread":{"thid":"22a77cd6-3131-4fd8-a2e8-b3a3a99b283b"}}
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).Handle -> INFO thread id value for the did exchange msg : 22a77cd6-3131-4fd8-a2e8-b3a3a99b283b
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).Handle -> INFO current state : requested
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).Handle -> INFO state will transition to -> completed if the msgType is processed
 [aries-framework/transport] 2019/10/03 14:51:16 UTC - http.processPOSTRequest -> ERROR incoming msg processing failed: invalid state transition: requested -> completed
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO next state that need to be executed immediately: noop
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO persisted the connection using 22a77cd6-3131-4fd8-a2e8-b3a3a99b283b and updating the state to responded
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO send post event for state responded
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).handle -> INFO --exited internal handle function--
 [aries-framework/did-exchange/service] 2019/10/03 14:51:16 UTC - didexchange.(*Service).startInternalListener.func1 -> ERROR processing of the message failed: failed to execute state completed handle inbound failed: failed to send msg using http outbound transport: received unsuccessful POST HTTP status from agent [http://localhost:36303, 500 Internal Server Error]```

### PR DESCRIPTION
**Pull request by fqutishat**
closes #408

Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

### CODE REVIEW

File: pkg/didcomm/protocol/didexchange/service.go [hunk 1]
```diff
@@ -142,6 +139,29 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
...
+		first = false
+		current, err = s.currentState(thid)
+		if err != nil {
+			return err
+		}
+		logger.Infof("current state : %s", current.Name())
+		if current.CanTransitionTo(next) {
+			return nil
+		}
+		logger.Infof("invalid state transition: %s -> %s [%d] retry",
```
Comment (Baha-sk) [inline pkg/didcomm/protocol/didexchange/service.go]: logger.Warnf maybe?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 2]
```diff
@@ -142,6 +139,29 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
 }
 
+func (s *Service) validateState(thid string, next state) error {
+	// TODO make it configurable #414
+	// for now we use defaults
+	r := retry.WithDefaults()
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: we need retries inside the state machine? why?

Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: we have race condition it's possible that you receive response before update the state @talwinder50 @llorllale

File: pkg/didcomm/protocol/didexchange/service.go (hunk 1 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: it's should be debug but for now will keep as info

**Pull request merged**